File: cat_vod/lib/src/utils/path.dart
/*
 * @Author: jadehh
 * @Date: 2024-08-14 15:37:24
 * @LastEditTime: 2024-08-20 19:01:33
 * @LastEditors: jadehh
 * @Description: 
 * @FilePath: \drama_source\cat_vod\lib\src\utils\path.dart
 * 
 */

import 'dart:io';
import 'dart:typed_data';
import 'package:cat_vod/src/utils/prefers.dart';
import 'package:cat_vod/src/utils/util.dart';
import 'package:drama_source_log/drama_source_log.dart';
import 'package:path/path.dart' as path;

class Path {
  static final String TAG = Path().runtimeType.toString();

  static String getApplicationDocumentsDirectory() {
    return Prefers.getString("application_documents_directory", defaultValue: "");
  }

  static String getApplicationCacheDirectory() {
    return Prefers.getString("application_cache_directory", defaultValue: "");
  }


  static _mkdir(Directory file, {String? name}) {
    try{
      if (!file.existsSync()) file.createSync();
    }catch (e,stackTrace){
      Log.e( "File Path:${file.absolute.path},${e.toString()}", stackTrace);
    }
    return file;
  }

  static bool exists(String path) {
    return Directory(path.replaceAll("file://", "")).existsSync();
  }

  static Directory root({String? child, String? name}) {
    if (child != null) return _mkdir(
        Directory(path.join((root()).absolute.path, child)), name: name);
    if (name != null) return Directory(path.join((root()).absolute.path, name));
    return Directory((getApplicationDocumentsDirectory()));
  }

  static FileSystemEntity download() {
    return _mkdir(Directory((getApplicationDocumentsDirectory()) + "Downloads"));
  }

  static FileSystemEntity cache({String? name}) {
    if (name != null) return File(path.join((cache()).absolute.path, name));
    return Directory((getApplicationCacheDirectory()));
  }

  static FileSystemEntity thunderCache() {
    FileSystemEntity internal = cache();
    String dir = Prefers.getString("thunder_cache_dir", defaultValue: internal.absolute.path);
    if (dir == internal.absolute.path) return internal;
    Directory _cache = Directory(dir);
    if (!_cache.existsSync()) return internal;
    return _cache;
  }

  static FileSystemEntity files({String? name}) {
    if (name != null) return File(path.join((files()).absolute.path, name));
    return Directory((getApplicationDocumentsDirectory()));
  }

  static String rootPath() {
    return (root()).absolute.path;
  }

  static String downloadPath() {
    return (download()).absolute.path;
  }

  static Directory database() {
    return _mkdir(Directory(path.join((root()).absolute.path, "databases")));
  }

  static Directory tv() {
    return _mkdir(Directory(path.join((root()).absolute.path, "TV")));
  }

  static Directory so() {
    return _mkdir(Directory(path.join((files()).absolute.path, "so")));
  }

  static FileSystemEntity js({String? name}) {
    if (name != null)
      return File(path.join((js()).absolute.path, name));
    return _mkdir(Directory(path.join((cache()).absolute.path, "js")));
  }

  static Directory py() {
    return _mkdir(Directory(path.join((cache()).absolute.path, "py")));
  }

  static Future<FileSystemEntity> jar({String? name}) async {
    if (name != null) return new File(
        path.join((await jar()).absolute.path, await Util.md5(name) + ".jar"));
    return _mkdir(Directory(path.join((cache()).absolute.path, "jar")));
  }

  static Directory doh() {
    return _mkdir(Directory(path.join((cache()).absolute.path, "doh")));
  }

  static Directory wall() {
    return _mkdir(Directory(path.join((cache()).absolute.path, "wall")));
  }

  static Directory exo() {
    return _mkdir(Directory(path.join((cache()).absolute.path, "exo")));
  }

  static Directory epg({String? name}) {
    if (name != null) return Directory(path.join((epg()).absolute.path, name));
    return _mkdir(Directory(path.join((cache()).absolute.path, "epg")));
  }

  static Directory jpa() {
    return _mkdir(Directory(path.join((cache()).absolute.path, "jpa")));
  }

  static Directory m3u8() {
    return _mkdir(Directory(path.join((cache()).absolute.path, "m3u8")));
  }

  static FileSystemEntity thunder({String? name}) {
    if (name != null)
      return _mkdir(new Directory(path.join((thunder()).absolute.path, name)));
    return _mkdir(Directory(path.join((cache()).absolute.path, "thunder")));
  }

  static Future<Directory> restore() async {
    return await _mkdir(
        Directory(path.join((await cache()).absolute.path, "restore")));
  }

  static File local(String path) {
    File file1 = File(path.replaceAll("file:/", ""));
    File file2 = File(path.replaceAll("file:/", rootPath()));
    return file2.existsSync() ? file2 : file1.existsSync() ? file1 : File(path);
  }

  static String read({String? path, File? file}) {
    File? _file;
    if (path != null) _file = local(path);
    if (file != null) _file = file;
    try {
      return _file!.readAsStringSync();
    } catch (e, stackTrace) {
      Log.e(e.toString(), stackTrace);
      return "";
    }
  }

  static List<int> readToByte(File file) {
    try {
      return file.readAsBytesSync();
    } catch (e, stackTrace) {
      Log.e(e.toString(), stackTrace);
      return [];
    }
  }

  static File write(FileSystemEntity fileSystemEntity, {List<int>? data, Uint8List? uint8List, String? str}) {
    File file = File(fileSystemEntity.absolute.path);
    file.createSync();
    try {
      if (data != null) file.writeAsBytesSync(data);
      if (uint8List != null) file.writeAsBytesSync(uint8List);
      if (str != null) {
        file.writeAsStringSync(str);
      }
      return file;
    } catch (e, stackTrace) {
      Log.e(e.toString(), stackTrace);
      return file;
    }
  }

  static File utf8(File file) {
    return write(cache(name: path.basename(file.path)), data: readToByte(file));
  }

  static void move(File inFile, File outFile) {
    copy(inFile: inFile, outFile: outFile);
    clear(file: inFile);
  }

  static copy({File? inFile, String? str, File? outFile}) {
    try {
      if (inFile != null && outFile != null) inFile.copySync(
          outFile.absolute.path);
      if (str != null && outFile != null) write(outFile, str: str);
    } catch (e, stackTrace) {
      Log.e(e.toString(), stackTrace);
    }
  }

  static List list(File dir) {
    List files = Directory(dir.path).listSync();
    return files;
  }

  static clear({File? file, FileSystemEntity? fileSystemEntity}) {
    try {
      if (file != null) file.deleteSync();
      if (fileSystemEntity != null) {
        fileSystemEntity.deleteSync(recursive: true);
        Directory(fileSystemEntity.absolute.path).createSync();
      }

    }
    catch (e, stackTrace) {
      Log.e(e.toString(), stackTrace);
    }
  }

// static File create(File file) {
// try {
// if (!file.canWrite()) file.setWritable(true);
// if (!file.exists()) file.createNewFile();
// Shell.exec("chmod 777 " + file);
// return file;
// } catch (Exception e) {
// e.printStackTrace();
// return file;
// }
// }
}

File: drama_source_app/android/catvod/src/main/java/com/github/catvod/utils/Path.java
package com.github.catvod.utils;

import android.os.Environment;
import android.util.Log;

import com.github.catvod.Init;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class Path {

    private static final String TAG = Path.class.getSimpleName();

    private static File mkdir(File file) {
        if (!file.exists()) file.mkdirs();
        return file;
    }

    public static boolean exists(String path) {
        return new File(path.replace("file://", "")).exists();
    }

    public static File root() {
        return Environment.getExternalStorageDirectory();
    }

    public static File download() {
        return Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS);
    }

    public static File cache() {
        return Init.context().getCacheDir();
    }

    public static File thunderCache() {
        File internal = Init.context().getCacheDir();
        String dir = Prefers.getString("thunder_cache_dir", internal.getAbsolutePath());
        if (dir.equals(internal.getAbsolutePath())) return internal;
        File cache = new File(dir);
        if (!cache.exists()) return internal;
        return cache;
    }

    public static File files() {
        return Init.context().getFilesDir();
    }

    public static String rootPath() {
        return root().getAbsolutePath();
    }

    public static String downloadPath() {
        return download().getAbsolutePath();
    }

    public static File tv() {
        return mkdir(new File(root() + File.separator + "TV"));
    }

    public static File so() {
        return mkdir(new File(files() + File.separator + "so"));
    }

    public static File js() {
        return mkdir(new File(cache() + File.separator + "js"));
    }

    public static File py() {
        return mkdir(new File(cache() + File.separator + "py"));
    }

    public static File jar() {
        return mkdir(new File(cache() + File.separator + "jar"));
    }

    public static File doh() {
        return mkdir(new File(cache() + File.separator + "doh"));
    }

    public static File exo() {
        return mkdir(new File(cache() + File.separator + "exo"));
    }

    public static File epg() {
        return mkdir(new File(cache() + File.separator + "epg"));
    }

    public static File jpa() {
        return mkdir(new File(cache() + File.separator + "jpa"));
    }

    public static File m3u8() {
        return mkdir(new File(cache() + File.separator + "m3u8"));
    }

    public static File thunder() {
        return mkdir(new File(thunderCache() + File.separator + "thunder"));
    }

    public static File restore() {
        return mkdir(new File(cache() + File.separator + "restore"));
    }

    public static File root(String name) {
        return new File(root(), name);
    }

    public static File root(String child, String name) {
        return new File(mkdir(new File(root(), child)), name);
    }

    public static File cache(String name) {
        return new File(cache(), name);
    }

    public static File files(String name) {
        return new File(files(), name);
    }

    public static File epg(String name) {
        return new File(epg(), name);
    }

    public static File js(String name) {
        return new File(js(), name);
    }

    public static File jar(String name) {
        return new File(jar(), Util.md5(name).concat(".jar"));
    }

    public static File thunder(String name) {
        return mkdir(new File(thunder(), name));
    }

    public static File local(String path) {
        File file1 = new File(path.replace("file:/", ""));
        File file2 = new File(path.replace("file:/", rootPath()));
        return file2.exists() ? file2 : file1.exists() ? file1 : new File(path);
    }

    public static String read(File file) {
        try {
            return read(new FileInputStream(file));
        } catch (Exception e) {
            return "";
        }
    }

    public static String read(String path) {
        try {
            return read(new FileInputStream(local(path)));
        } catch (Exception e) {
            return "";
        }
    }

    public static String read(InputStream is) {
        try {
            byte[] data = new byte[is.available()];
            is.read(data);
            is.close();
            return new String(data, StandardCharsets.UTF_8);
        } catch (IOException e) {
            e.printStackTrace();
            return "";
        }
    }

    public static byte[] readToByte(File file) {
        try {
            FileInputStream is = new FileInputStream(file);
            byte[] data = new byte[is.available()];
            is.read(data);
            is.close();
            return data;
        } catch (IOException e) {
            e.printStackTrace();
            return new byte[0];
        }
    }

    public static File write(File file, byte[] data) {
        try {
            FileOutputStream fos = new FileOutputStream(create(file));
            fos.write(data);
            fos.flush();
            fos.close();
            return file;
        } catch (Exception ignored) {
            ignored.printStackTrace();
            return file;
        }
    }

    public static File utf8(File file) {
        return write(cache(file.getName()), Util.utf8(readToByte(file)));
    }

    public static void move(File in, File out) {
        copy(in, out);
        clear(in);
    }

    public static void copy(File in, File out) {
        try {
            copy(new FileInputStream(in), out);
        } catch (Exception ignored) {
        }
    }

    public static void copy(InputStream in, File out) {
        try {
            int read;
            byte[] buffer = new byte[8192];
            FileOutputStream fos = new FileOutputStream(create(out));
            while ((read = in.read(buffer)) != -1) fos.write(buffer, 0, read);
            fos.close();
            in.close();
        } catch (Exception ignored) {
        }
    }

    public static List<File> list(File dir) {
        File[] files = dir.listFiles();
        return files == null ? Collections.emptyList() : Arrays.asList(files);
    }

    public static void clear(File dir) {
        if (dir == null) return;
        if (dir.isDirectory()) for (File file : list(dir)) clear(file);
        if (dir.delete()) Log.d(TAG, "Deleted:" + dir.getAbsolutePath());
    }

    public static File create(File file) throws Exception {
        try {
            if (!file.canWrite()) file.setWritable(true);
            if (!file.exists()) file.createNewFile();
            Shell.exec("chmod 777 " + file);
            return file;
        } catch (Exception e) {
            e.printStackTrace();
            return file;
        }
    }
}
